### PR TITLE
Mc refactor

### PR DESCRIPTION
**Pull request by hellkite500**
## Issue Addressed

Currently, `dmc.py` contains both the MC routing math AND the torch NN module in a single class and requires a lot of data passing and handling via the torch forward/backward interfaces.

This poses a couple issues, but the most direct one is a hard time running a MC model "standalone" and with different parameters from different sources.

This also makes maintaining, testing, and extending harder given how tightly coupled the torch and MC code is right now.

## Description

The proposed changes here separate the mathematical core and data requirements for MC into its own module and class, and then interface with that via the TorchMC class.

A comprehensive migration test suite is included to validate the new interface and backwards compatibility with dmc.

These tests and refactor need careful review.  The next steps once this is verified is refactoring the data components in the new mmc module to make them more module and easier to construct, validate, and pass through the system.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code cleanup/refactor
- [ ] Documentation update

Other (please specify):

## Checklist

- [x] Branch is up to date with master
- [x] Updated tests or added new tests
- [x] Tests & pre-commit hooks pass
- [ ] Updated documentation (if applicable)
- [x] Code follows established style and conventions

# Muskingum-Cunge Refactoring Summary

Splitting `dmc.py` into two modules, one for differentiable parameter leaning, one for running MC.

### 1. `src/ddr/routing/mmc.py` - Core Mathematical Implementation

**Purpose**: Contains the pure mathematical implementation of the Muskingum-Cunge routing algorithm.

**Key Features**:
- `MuskingumCunge` class that manages all hydrofabric data, parameters, and routing calculations
- Self-contained implementation that handles data setup and management internally
- Clean separation of mathematical logic from PyTorch neural network concerns
- Comprehensive input validation and parameter management

**Main Methods**:
- `setup_inputs()`: Configures all inputs (hydrofabric, streamflow, spatial parameters)
- `forward()`: Performs complete routing calculation through time series
- `route_timestep()`: Routes flow for a single timestep
- `calculate_muskingum_coefficients()`: Computes routing coefficients
- Utility methods for sparse matrix operations

### 2. `src/ddr/routing/torch_mc.py` - PyTorch Neural Network Module

**Purpose**: Provides a PyTorch `nn.Module` wrapper around the core implementation for training and inference.

**Key Features**:
- `TorchMC` class that inherits from `torch.nn.Module`
- GPU/CPU device management with proper tensor movement
- Full compatibility with PyTorch training pipelines
- Backward compatibility with the original `dmc` interface
- Automatic differentiation support for training

**Main Methods**:
- `forward()`: PyTorch forward pass using the core implementation
- `to()`, `cuda()`, `cpu()`: Device management methods
- `state_dict()`, `load_state_dict()`: Model saving/loading
- Compatibility methods that delegate to the core implementation

### 1. **Separation of Concerns**
- Mathematical routing logic separated from PyTorch-specific functionality
- Core implementation is framework-agnostic and can be reused
- Neural network wrapper provides PyTorch integration

### 2. **Data Management**
- `MuskingumCunge` class now manages all input data internally
- Eliminates need to pass hydrofabric data through multiple method calls
- Cleaner, more maintainable interface

### 3. **Backward Compatibility**
- `TorchMC` is aliased as `dmc` for drop-in replacement capability
- Existing training and evaluation scripts require no changes
- All original method signatures preserved through compatibility methods

### 4. **Device Flexibility**
- Proper PyTorch device management with `to()`, `cuda()`, `cpu()` methods
- Tensors automatically moved to correct devices
- GPU/CPU training fully supported

### 5. **Code Organization**
- Better structured codebase with clear responsibilities
- Easier to test and maintain individual components
- More modular architecture supports future extensions

### Using the Core Implementation
```python
from ddr.routing.mmc import MuskingumCunge

# Create routing engine
mc = MuskingumCunge(cfg, device='cuda')

# Setup all inputs at once
mc.setup_inputs(hydrofabric, streamflow, spatial_parameters)

# Run routing
output = mc.forward()
```

### Using the PyTorch Module
```python
from ddr.routing.torch_mc import TorchMC

# Create PyTorch module
model = TorchMC(cfg, device='cuda')

# Use like any PyTorch module
kwargs = {
    'hydrofabric': hydrofabric,
    'streamflow': streamflow,
    'spatial_parameters': spatial_parameters
}
output = model(**kwargs)

# Training compatibility
loss = criterion(output['runoff'], targets)
loss.backward()
```

### Drop-in Replacement
```python
# Original code - no changes needed
from ddr.routing.dmc import dmc

routing_model = dmc(cfg=cfg, device=cfg.device)
output = routing_model(**dmc_kwargs)
```

### For Existing Code
- **No changes required** - the `TorchMC` class provides full backward compatibility
- Import statements continue to work: `from ddr.routing.dmc import dmc`
- All existing method calls and interfaces preserved

### For New Development
- Use `MuskingumCunge` directly for pure routing calculations
- Use `TorchMC` for PyTorch training and inference
- Leverage the improved data management for cleaner code

## Benefits

1. **Maintainability**: Clear separation of concerns makes code easier to understand and maintain
2. **Testability**: Individual components can be tested in isolation
3. **Reusability**: Core routing logic can be used in different contexts
4. **Extensibility**: Modular design supports future enhancements
5. **Performance**: More efficient data management reduces overhead
6. **Compatibility**: Zero-impact migration for existing code

The refactoring successfully achieves the goal of creating a more modular, maintainable, and flexible Muskingum-Cunge routing implementation while preserving full backward compatibility.

## Testing

# Comprehensive Test Suite Summary

## Overview
This document provides a comprehensive overview of the pytest unit test suite created for the Muskingum-Cunge routing implementation, covering the original `dmc.py` and the refactored `mmc.py` and `torch_mc.py` classes.

### 1. Test Utilities (`tests/routing/test_utils.py`)
- **Mock Configuration**: Creates test configurations with realistic parameter ranges
- **Mock Hydrofabric**: Generates realistic hydrofabric objects with network topology
- **Mock Streamflow**: Creates test streamflow data with temporal variation
- **Mock Spatial Parameters**: Generates spatial parameter data for testing
- **Test Scenarios**: Provides parameterized test scenarios for different network sizes
- **Assertion Helpers**: Utility functions for validating tensor properties and outputs

### 2. Original DMC Tests (`tests/routing/test_dmc.py`)
**Total Tests: 27**

#### Utility Functions (3 tests)
- ✅ `test_log_base_q`: Tests logarithm utility function
- ✅ `test_get_trapezoid_velocity`: Tests Manning's equation velocity calculation
- ✅ `test_get_trapezoid_velocity_edge_cases`: Tests edge cases for velocity calculation

#### Initialization (4 tests)
- ✅ `test_init_cpu`: CPU device initialization
- ✅ `test_init_default_device`: Default device handling
- ✅ `test_init_none_device`: None device parameter handling
- ✅ `test_parameter_bounds_setup`: Parameter bounds configuration

#### Sparse Operations (3 tests)
- ✅ `test_sparse_eye`: Sparse identity matrix creation
- ✅ `test_sparse_diag`: Sparse diagonal matrix creation
- ✅ `test_fill_op`: Fill operation for routing matrix

#### Forward Pass (3 tests)
- ✅ `test_forward_basic`: Basic forward pass functionality
- ✅ `test_forward_parameter_setup`: Parameter setup during forward pass
- ✅ `test_forward_discharge_clamping`: Discharge value clamping

#### Error Handling (1 test)
- ✅ `test_cuda_out_of_memory_handling`: CUDA memory error handling

#### Integration (5 tests)
- ✅ `test_different_network_sizes`: Multiple network size scenarios
- ✅ `test_reproducibility`: Deterministic behavior validation

#### PyTorch Integration (2 tests)
- ✅ `test_is_nn_module`: PyTorch module inheritance
- ✅ `test_gradient_flow`: Gradient computation and backpropagation

#### State Management (2 tests)
- ✅ `test_discharge_state_updates`: Discharge state tracking
- ✅ `test_network_assignment`: Network matrix assignment

#### Parameter Training (4 tests)
- ✅ `test_parameter_training`: Gradient flow validation for neural network parameter training across multiple scenarios

### 3. Refactored MMC Tests (`tests/routing/test_mmc.py`)
**Total Tests: 26**

#### Core Class Tests
- ✅ **Initialization** (3 tests): CPU/default device setup, parameter bounds
- ✅ **Progress Tracking** (1 test): Epoch and mini-batch tracking
- ✅ **Input Setup** (3 tests): Hydrofabric, streamflow, and parameter setup
- ✅ **Sparse Operations** (3 tests): Identity, diagonal, and fill operations
- ✅ **Coefficient Calculation** (2 tests): Muskingum coefficients with edge cases
- ✅ **Pattern Mapper** (1 test): Sparse pattern mapping functionality
- ✅ **Route Timestep** (2 tests): Single timestep routing with clamping
- ✅ **Forward Pass** (3 tests): Complete forward pass with state updates
- ✅ **Integration** (6 tests): Multiple scenarios and full workflow
- ✅ **Error Handling** (2 tests): Input validation and invalid state handling

### 4. PyTorch Wrapper Tests (`tests/routing/test_torch_mc.py`)
**Total Tests: 37**

#### Core Module Tests
- ✅ **Initialization** (4 tests): Device setup and routing engine configuration
- ✅ **Device Management** (6 tests): CPU/CUDA movement and device handling
- ✅ **Progress Tracking** (1 test): Progress information propagation
- ✅ **Forward Pass** (4 tests): PyTorch module integration and device handling
- ✅ **Compatibility Methods** (4 tests): Delegation to routing engine
- ✅ **State Management** (3 tests): PyTorch state dict handling
- ✅ **PyTorch Integration** (3 tests): Module behavior and gradient flow
- ✅ **Integration** (5 tests): End-to-end workflows and network scenarios
- ✅ **Backward Compatibility** (3 tests): Drop-in replacement validation

#### Parameter Training (4 tests)
- ✅ `test_parameter_training`: Comprehensive gradient flow validation for neural network parameter training with gradient retention and unused parameter filtering

### 5. Validation Tests (`tests/routing/test_validation.py`)
**Total Tests: 14**

#### Refactored vs Original Validation (6 tests)
- ✅ `test_initialization_parity`: Identical initialization behavior
- ✅ `test_sparse_operations_parity`: Identical sparse matrix operations
- ✅ `test_fill_op_parity`: Identical fill operation results
- ✅ `test_forward_pass_interface_parity`: Identical forward pass interfaces
- ✅ `test_parameter_setup_parity`: Identical parameter handling
- ✅ `test_state_management_parity`: Identical state management

#### Core Validation (2 tests)
- ✅ `test_core_setup_and_forward`: Core routing engine functionality
- ✅ `test_core_encapsulation`: Proper data encapsulation

#### Backward Compatibility (4 tests)
- ✅ `test_training_script_compatibility`: Training script compatibility
- ✅ `test_evaluation_script_compatibility`: Evaluation script compatibility
- ✅ `test_import_compatibility`: Import statement compatibility
- ✅ `test_device_management_compatibility`: Device management compatibility

#### Performance Validation (2 tests)
- ✅ `test_memory_usage_comparable`: Memory usage comparison
- ✅ `test_forward_pass_efficiency`: Performance comparison

### Key Testing Areas Covered:
1. **Functional Correctness**: All mathematical operations and routing algorithms
2. **Device Compatibility**: CPU and CUDA device handling
3. **PyTorch Integration**: Module behavior, gradients, and state management
4. **Gradient Flow**: Neural network parameter training and backpropagation validation
5. **Error Handling**: Input validation and edge case handling
6. **Performance**: Memory usage and computational efficiency
7. **Backward Compatibility**: Drop-in replacement validation
8. **Integration**: End-to-end workflows and real-world scenarios

### Test Execution Results:
- **Total Tests**: 104 tests across all modules  
- **Success Rate**: 98.1% (102/104 tests passing, 2 skipped)
- **Test Types**: Unit, Integration, Validation, Performance, Gradient Flow
- **Code Coverage**: Comprehensive coverage of all public APIs and core functionality

### Key Testing Strategies Used:
1. **Mocking**: Extensive use of mocks for complex dependencies
2. **Parameterization**: Multiple test scenarios with different network sizes
3. **Property-Based Testing**: Tensor property validation utilities
4. **Regression Testing**: Validation against original implementation
5. **Edge Case Testing**: Boundary conditions and error states
6. **Performance Testing**: Memory and timing comparisons
7. **Gradient Flow Testing**: Neural network parameter training validation with gradient retention
8. **Unused Parameter Filtering**: Smart filtering of spatial parameters not used in routing computation

#### Gradient Flow Validation
Added comprehensive gradient flow testing to ensure neural network parameter training works correctly with the routing models:

**Key Issues Resolved:**
- **Non-leaf Tensor Gradients**: Implemented `.retain_grad()` calls for non-leaf tensors (neural network outputs) during gradient testing
- **Unused Parameter Filtering**: Added smart filtering to only check gradients on spatial parameters actually used by the routing engine (based on `cfg.params.parameter_ranges.range`)
- **Gradient Connection Verification**: Ensured proper gradient flow from neural network outputs through denormalization to final routing computation

**Technical Details:**
- Fixed gradient retention for `spatial_parameters['n']` and `spatial_parameters['q_spatial']` 
- Filtered out unused `spatial_parameters['p_spatial']` parameter that wasn't included in routing computation
- Validated gradient flow across multiple test scenarios with different network sizes

To run all tests:
```bash
# All routing tests
pytest tests/routing/ -v


# Individual test modules
pytest tests/routing/test_dmc.py -v
pytest tests/routing/test_mmc.py -v
pytest tests/routing/test_torch_mc.py -v
pytest tests/routing/test_validation.py -v
```

## Continuous Integration Ready

The test suite is designed to be CI/CD ready with:
- Fast execution (< 3 seconds total)
- No external dependencies beyond PyTorch
- Comprehensive error reporting
- Deterministic behavior
- Cross-platform compatibility
- Robust gradient flow validation for neural network training scenarios

## Conclusion

This comprehensive test suite ensures:
1. **Correctness**: Mathematical operations are accurate
2. **Reliability**: Code behaves consistently across scenarios
3. **Maintainability**: Changes can be validated against existing behavior
4. **Performance**: No regression in computational efficiency
5. **Compatibility**: Seamless integration with existing codebase
6. **Gradient Flow**: Neural network parameter training works correctly with proper gradient retention and backpropagation

The test suite provides confidence that the refactored Muskingum-Cunge implementation maintains full compatibility with the original while providing improved modularity, maintainability, and robust support for neural network-based parameter training.

### CODE REVIEW

**Comment by taddyb** [comment]
Something that needs to be tested is if this code can train and there are no graph breaks or NaN grad operations introduced

**Comment by hellkite500** [comment]
What do you mean by "graph breaks"? I can put toghter a small training test based on the mocks, but I'm not sure that is sufficient what you are talking about here given I'm not sure exactly what you are talking about here 😄.

**Comment by taddyb** [comment]
> What do you mean by "graph breaks"? 

I forgot for a minute this is a graph-based project lol. We have to ensure the gradient is attached to tensors from the NN runs -> dMC -> the loss calculation. This means:
- no in place operations as you'll get this error
```ValueError: can’t optimize a non-leaf Tensor```
- no graph breaks. This means the _computational graph_ connecting the NN weights to the loss function. So no `.detach()` or a new creation of a tensor
- No math with causes NaN values
  - examples include division by 0, raising a power by 0 or a negative number, problems with clamps

Hopefully this is a good start

**Comment by hellkite500** [comment]
For some added context, I ran these gradient tests manually after using both these lines in the `backward()` function of the sparse solver

```python
        return (None,)*7
        return ( torch.Tensor([np.nan]*9), None, None, torch.tensor([np.nan]*5), None, None, None)
```
and got the anticipated assertions from the test (running just scenario0 for the second case so I could hard code the right shapes for the NaN gradient injection...)

File: TEST_SUMMARY.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+4. **Regression Testing**: Validation against original implementation
+5. **Edge Case Testing**: Boundary conditions and error states
+6. **Performance Testing**: Memory and timing comparisons
+
+## Test Execution
+
+To run all tests:
+```bash
+# All routing tests
+pytest tests/routing/ -v
```
Comment (taddyb) [inline TEST_SUMMARY.md]: FYI, can also be run through 
```sh
uv run pytest     
```

I think I'm too much of a uv fan boy

**Review by taddyb — COMMENTED** [review]
I want to look at this once more tomorrow to make sure I am fully caught up with all of the changes, but I really like what I'm seeing !

File: tests/routing/gradient_utils.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 @version 0.1
 
 A set of utility functions to find tensors in an object that may have gradients.
+An object can be a dataclass, torch.nn.ModuleList, list, tuple, etc.
```
Comment (hellkite500) [inline tests/routing/gradient_utils.py]: Perhaps a more percise definition is

Any of the following container objects:
`list`, `tuple`, `torch.nn.ModuleList`, or `dict`
Or any object containing a `__dict__` attribute

**Review by taddyb — APPROVED** [review]
Did another pass on the tests. I think this is ready for merging. Approved!

**Comment by hellkite500** [comment]
Should we fix the ruff check?

**Comment by taddyb** [comment]
> Should we fix the ruff check?

My fork didn't have the pre-commit hooks installed. Linting should pass now
